File: backend/src/main/java/com/dopplertask/doppler/domain/action/common/XMLAction.java
package com.dopplertask.doppler.domain.action.common;

import com.dopplertask.doppler.domain.ActionResult;
import com.dopplertask.doppler.domain.StatusCode;
import com.dopplertask.doppler.domain.TaskExecution;
import com.dopplertask.doppler.domain.action.Action;
import com.dopplertask.doppler.service.BroadcastListener;
import com.dopplertask.doppler.service.TaskService;
import com.dopplertask.doppler.service.VariableExtractorUtil;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.xml.XmlMapper;

import java.io.IOException;
import java.util.List;

import javax.persistence.Column;
import javax.persistence.DiscriminatorValue;
import javax.persistence.Entity;
import javax.persistence.Lob;
import javax.persistence.Table;

@Entity
@Table(name = "XMLAction")
@DiscriminatorValue("xml_action")
public class XMLAction extends Action {

    @Lob
    @Column(columnDefinition = "TEXT")
    private String content;

    @Column
    private XMLActionType type = XMLActionType.JSON_TO_XML;

    public XMLAction() {
    }


    @Override
    public ActionResult run(TaskService taskService, TaskExecution execution, VariableExtractorUtil variableExtractorUtil, BroadcastListener broadcastListener) throws IOException {
        String contentVariable = variableExtractorUtil.extract(content, execution, getScriptLanguage());

        ActionResult actionResult = new ActionResult();

        if (contentVariable != null && !contentVariable.isEmpty()) {
            if (type == XMLActionType.XML_TO_JSON) {
                XmlMapper xmlMapper = new XmlMapper();
                JsonNode node = xmlMapper.readTree(contentVariable.getBytes());
                ObjectMapper jsonMapper = new ObjectMapper();

                actionResult.setOutput(jsonMapper.writeValueAsString(node));

            } else if (type == XMLActionType.JSON_TO_XML) {
                ObjectMapper objectMapper = new ObjectMapper();
                XmlMapper xmlMapper = new XmlMapper();
                JsonNode tree = objectMapper.readTree(content);

                actionResult.setOutput(xmlMapper.writer().withRootName("xml").writeValueAsString(tree));

            }

            actionResult.setStatusCode(StatusCode.SUCCESS);
        } else {
            actionResult.setErrorMsg("No output");
            actionResult.setStatusCode(StatusCode.FAILURE);
        }
        return actionResult;
    }

    public String getContent() {
        return content;
    }

    public void setContent(String content) {
        this.content = content;
    }

    @Override
    public List<PropertyInformation> getActionInfo() {
        List<PropertyInformation> actionInfo = super.actionInfo;

        actionInfo.add(new PropertyInformation("content", "Content", PropertyInformation.PropertyInformationType.MULTILINE));
        actionInfo.add(new PropertyInformation("type", "Mode", PropertyInformation.PropertyInformationType.DROPDOWN, "JSON_TO_XML", "Convert from and to XML", List.of(
                new PropertyInformation("JSON_TO_XML", "JSON to XML"),
                new PropertyInformation("XML_TO_JSON", "XML to JSON")

        )));
        return actionInfo;
    }

    @Override
    public String getDescription() {
        return "Converts from/to XML";
    }

    public XMLActionType getType() {
        return type;
    }

    public void setType(XMLActionType type) {
        this.type = type;
    }
}

File: backend/src/main/java/com/dopplertask/doppler/domain/action/common/XMLAction.kt
package com.dopplertask.doppler.domain.action.common

import com.dopplertask.doppler.domain.ActionResult
import com.dopplertask.doppler.domain.StatusCode
import com.dopplertask.doppler.domain.TaskExecution
import com.dopplertask.doppler.domain.action.Action
import com.dopplertask.doppler.domain.action.Action.PropertyInformation.PropertyInformationType
import com.dopplertask.doppler.service.BroadcastListener
import com.dopplertask.doppler.service.TaskService
import com.dopplertask.doppler.service.VariableExtractorUtil
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.xml.XmlMapper
import java.io.IOException
import javax.persistence.*

@Entity
@Table(name = "XMLAction")
@DiscriminatorValue("xml_action")
class XMLAction : Action() {
    @Lob
    @Column(columnDefinition = "TEXT")
    var content: String? = null

    @Column
    var type = XMLActionType.JSON_TO_XML

    @Throws(IOException::class)
    override fun run(taskService: TaskService, execution: TaskExecution, variableExtractorUtil: VariableExtractorUtil, broadcastListener: BroadcastListener?): ActionResult {
        val contentVariable = variableExtractorUtil.extract(content, execution, scriptLanguage)
        val actionResult = ActionResult()
        if (contentVariable != null && !contentVariable.isEmpty()) {
            if (type == XMLActionType.XML_TO_JSON) {
                val xmlMapper = XmlMapper()
                val node = xmlMapper.readTree(contentVariable.toByteArray())
                val jsonMapper = ObjectMapper()
                actionResult.output = jsonMapper.writeValueAsString(node)
            } else if (type == XMLActionType.JSON_TO_XML) {
                val objectMapper = ObjectMapper()
                val xmlMapper = XmlMapper()
                val tree = objectMapper.readTree(content)
                actionResult.output = xmlMapper.writer().withRootName("xml").writeValueAsString(tree)
            }
            actionResult.statusCode = StatusCode.SUCCESS
        } else {
            actionResult.errorMsg = "No output"
            actionResult.statusCode = StatusCode.FAILURE
        }
        return actionResult
    }

    override val actionInfo: MutableList<PropertyInformation>
        get() {
            val actionInfo = super.actionInfo
            actionInfo.add(PropertyInformation("content", "Content", PropertyInformationType.MULTILINE))
            actionInfo.add(PropertyInformation("type", "Mode", PropertyInformationType.DROPDOWN, "JSON_TO_XML", "Convert from and to XML", listOf(
                    PropertyInformation("JSON_TO_XML", "JSON to XML"),
                    PropertyInformation("XML_TO_JSON", "XML to JSON")
            )))
            return actionInfo
        }

    override val description: String
        get() = "Converts from/to XML"

}